README
README

This project implements the least-common ancestor algorithm using
segment trees, which is loglinear in space and time to construct and
linear in query time. The application accepts as parameters the Euler
Circuit and corresponding auxiliary data structures including the
levels of the corresponding vertices and the first position of a
vertex in the Euler Circuit. These arrays can be computed using the
Euler-Circuit application also on my github [1]. Additionally, a query
set has to be defined for which the LCA is computed pair-wise.


CONFIGURATION

The code is implemented in C++ using the boost libraries, in
particular the Boost Graph Library, Filesystem, and Program Options.

OpenMP can be enabled using the --enable-openmp flag at the
configuration step.


EXECUTION

General Configuration:
  --help                produce help message
  --version             show the version

I/O Configuration:
  --results arg (=./results) results directory.
  --euler_circuit arg        Filename of the Euler Circuit.
  --euler_levels arg         Filename of the vertex levels in the Euler 
                             Circuit.
  --euler_positions arg      Filename of the vertex positions in the Euler 
                             Circuit.
  --query_set arg            Filename of the query set to compute pair-wise 
                             LCAs over.

[1] https://github.com/dahlem/Euler-Circuit


ChangeLog
* [0.5] - 2015-05-05
** Changed
 - Implemented LCA using the euler circuit and level arrays

* [0.6] - 2015-05-08
** Fixed
 - fixed the RMQ query


src/main/RMQ.cc
// Copyright (C) 2015 Dominik Dahlem <Dominik.Dahlem@gmail.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

/** @file RMQ.cc
 * Implementation of the RMQ routines
 *
 * @author Dominik Dahlem
 */
#if HAVE_CONFIG_H
# include <config.h>
#endif

#ifndef __STDC_CONSTANT_MACROS
# define __STDC_CONSTANT_MACROS
#endif /* __STDC_CONSTANT_MACROS */

#ifndef NDEBUG
# include <algorithm>
# include <iostream>
# include <iterator>
#endif /* NDEBUG */

#include <cmath>
#include <cstring>

#include <boost/cstdint.hpp>
#include <boost/numeric/conversion/cast.hpp>

#include "RMQ.hh"

namespace rmq
{

RMQ::RMQ() {}

RMQ::~RMQ()
{
  if (m_M) { delete [] m_M; }
}

void RMQ::initialise(itypes::IntVector p_A)
    throw (RMQException)
{
  try {
    boost::uint32_t N = boost::numeric_cast<boost::uint32_t>(p_A.size());
    boost::uint32_t x = boost::numeric_cast<boost::uint32_t>(std::ceil(log2(N))); // height of segment tree
    boost::uint32_t size = 2 * std::pow(2, x) - 1; // max size of segment tree
    m_M = new boost::int32_t[size];
    std::memset(m_M, -1, size * sizeof(boost::int32_t));

#ifndef NDEBUG
    std::cout << "Initialise M: " << std::endl;
    std::copy(m_M, m_M+size, std::ostream_iterator<boost::int32_t>(std::cout, "\t"));
    std::cout << std::endl;
#endif /* NDEBUG */

    create(p_A, 1, 0, N-1);
  } catch(boost::numeric::negative_overflow& e) {
    throw RMQException(RMQException::BAD_CAST);
  } catch(boost::numeric::positive_overflow& e) {
    throw RMQException(RMQException::BAD_CAST);
  }
}

boost::int32_t RMQ::query(itypes::IntVector p_A, boost::uint32_t p_node, boost::uint32_t p_lb,
                           boost::uint32_t p_ub, boost::uint32_t p_i, boost::uint32_t p_j)
{
  boost::int32_t p1, p2;

#ifndef NDEBUG
  itypes::IntVector::iterator b = p_A.begin(), e = p_A.begin();
  std::advance(b, p_lb);
  std::advance(e, p_ub+1);
  std::cout << "Current node: " << p_node << "=" << m_M[p_node] << std::endl;
  std::cout << "Query range [" << p_i << ", " << p_j << "] within array [" << p_lb << ", " << p_ub << "]: ";
  std::copy(b, e, std::ostream_iterator<boost::uint32_t>(std::cout, "\t"));
  std::cout << std::endl;
#endif /* NDEBUG */

  // query interval does not intersect
  if (p_i > p_ub || p_j < p_lb) {
#ifndef NDEBUG
    std::cout << "Query interval does not intersect" << std::endl;
#endif /* NDEBUG */
    return -1;
  }

  // the current interval is included in the query interval
  if (p_i <= p_lb && p_j >= p_ub) {
#ifndef NDEBUG
    std::cout << "The current interval is included in the query interval" << std::endl;
#endif /* NDEBUG */
    return m_M[p_node];
  }

  // compute the minimum position in the
  // left and right part of the interval
  boost::uint32_t midPoint = (p_lb + p_ub) / 2;
  p1 = query(p_A, 2 * p_node, p_lb, midPoint, p_i, p_j);
  p2 = query(p_A, 2 * p_node + 1, midPoint + 1, p_ub, p_i, p_j);

#ifndef NDEBUG
  std::cout << "p1=" << p1 << ", p2=" << p2 << std::endl;
#endif /* NDEBUG */

  // return minimum
  if (p1 == -1) { return p2; }
  if (p2 == -1) { return p1; }
  if (p_A[p1] <= p_A[p2]) {return p1; }

  return p2;
}

void RMQ::create(itypes::IntVector p_A, boost::uint32_t p_node, boost::uint32_t p_lb, boost::uint32_t p_ub)
{
  if (p_lb == p_ub) {
    m_M[p_node] = p_lb;
  } else {
    // create the left and right subtrees
    boost::uint32_t midPoint = (p_lb + p_ub) / 2;
    create(p_A, 2 * p_node, p_lb, midPoint);
    create(p_A, 2 * p_node + 1, midPoint + 1, p_ub);

    // search for the minimum value in the first and
    // second half of the interval
    if (p_A[m_M[2 * p_node]] <= p_A[m_M[2 * p_node + 1]]) {
      m_M[p_node] = m_M[2 * p_node];
    } else {
      m_M[p_node] = m_M[2 * p_node + 1];
    }
  }
}


}


src/main/main.cc
// Copyright (C) 2015 Dominik Dahlem <Dominik.Dahlem@gmail.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

#if HAVE_CONFIG_H
# include <config.h>
#endif

#ifndef __STDC_CONSTANT_MACROS
# define __STDC_CONSTANT_MACROS
#endif /* __STDC_CONSTANT_MACROS */

#ifndef NDEBUG
# include <algorithm>
# include <iterator>
#endif /* NDEBUG */

#include <fstream>
#include <iostream>
#include <vector>

#include <boost/cstdint.hpp>
#include <boost/lexical_cast.hpp>
#include <boost/tokenizer.hpp>

#include <boost/algorithm/string/split.hpp>
#include <boost/algorithm/string/trim.hpp>

#include "CL.hh"

#include "Types.hh"
namespace itypes = types;

#include "RMQ.hh"
namespace irmq = rmq;

typedef boost::tokenizer <boost::escaped_list_separator <char> > Tokenizer;


int main(int argc, char *argv[])
{
  args_t args;
  CL cl;

  if (cl.parse(argc, argv, args)) {
    return EXIT_SUCCESS;
  }

  // parse the euler circuit
  std::ifstream eulerCircuitFile;
  eulerCircuitFile.open(args.euler_circuit.c_str());
  if (!eulerCircuitFile.is_open()) {
    std::cerr << "Could not open file: " << args.euler_circuit << std::endl;
    return EXIT_FAILURE;
  }

#ifndef NDEBUG
  std::cout << "Reading the Euler Circuit..." << std::endl;
#endif /* NDEBUG */

  itypes::StringVector euler_circuit;
  std::string line;
  while (!eulerCircuitFile.eof()) {
    std::getline(eulerCircuitFile, line);
#ifndef NDEBUG
    std::cout << "Read line: " << line << std::endl;
#endif /* NDEBUG */

    if (line != "") {
      boost::algorithm::trim(line);
      euler_circuit.push_back(line);
    }
  }
  eulerCircuitFile.close();

  // parse the euler levels
  std::ifstream eulerLevelsFile;
  eulerLevelsFile.open(args.euler_levels.c_str());
  if (!eulerLevelsFile.is_open()) {
    std::cerr << "Could not open file: " << args.euler_levels << std::endl;
    return EXIT_FAILURE;
  }

#ifndef NDEBUG
  std::cout << "Reading the Euler Levels..." << std::endl;
#endif /* NDEBUG */

  itypes::IntVector euler_levels;
  while (!eulerLevelsFile.eof()) {
    std::getline(eulerLevelsFile, line);
#ifndef NDEBUG
    std::cout << "Read line: " << line << std::endl;
#endif /* NDEBUG */

    if (line != "") {
      boost::algorithm::trim(line);
      euler_levels.push_back(boost::lexical_cast<boost::uint32_t>(line));
    }
  }
  eulerLevelsFile.close();


  // parse the euler positions
  std::ifstream eulerPositionsFile;
  eulerPositionsFile.open(args.euler_positions.c_str());
  if (!eulerPositionsFile.is_open()) {
    std::cerr << "Could not open file: " << args.euler_positions << std::endl;
    return EXIT_FAILURE;
  }

#ifndef NDEBUG
  std::cout << "Reading the Euler Positions..." << std::endl;
#endif /* NDEBUG */

  itypes::StringIntMap euler_positions;
  while (!eulerPositionsFile.eof()) {
    std::getline(eulerPositionsFile, line);
#ifndef NDEBUG
    std::cout << "Read line: " << line << std::endl;
#endif /* NDEBUG */

    if (line != "") {
      itypes::StringVector tokens;
      boost::split(tokens, line, boost::is_any_of(","), boost::token_compress_on);
      if (tokens.size() != 2) {
        eulerPositionsFile.close();
        std::cerr << "Each line of the euler positions file can only contain two tokens: <key>,<value>!" << std::endl;
        return EXIT_FAILURE;
      } else {
        if (euler_positions.find(tokens[0]) != euler_positions.end()) {
          eulerPositionsFile.close();
          std::cerr << "The Euler Positions cannot contain duplicates: " << tokens[0] << std::endl;
          return EXIT_FAILURE;
        } else {
          euler_positions[tokens[0]] = boost::lexical_cast<boost::uint32_t>(tokens[1]);
        }
      }
    }
  }
  eulerPositionsFile.close();

#ifndef NDEBUG
  std::cout << "Euler Circuit: ";
  std::copy(euler_circuit.begin(), euler_circuit.end(), std::ostream_iterator<std::string>(std::cout, "\t"));
  std::cout << std::endl << "Euler Levels:  ";
  std::copy(euler_levels.begin(), euler_levels.end(), std::ostream_iterator<boost::uint32_t>(std::cout, "\t"));
  std::cout << std::endl;
#endif /* NDEBUG */

  std::ifstream verticesFile;
  verticesFile.open(args.query_set.c_str());
  if (!verticesFile.is_open()) {
    std::cerr << "Could not open file: " << args.query_set << std::endl;
    return EXIT_FAILURE;
  }

#ifndef NDEBUG
  std::cout << "Reading the vertices to compute LCAs over..." << std::endl;
#endif /* NDEBUG */

  itypes::StringVector queries;
  while (!verticesFile.eof()) {
    std::getline(verticesFile, line);
#ifndef NDEBUG
    std::cout << "Read line: " << line << std::endl;
#endif /* NDEBUG */

    if (line != "") {
      boost::algorithm::trim(line);
      queries.push_back(line);
    }
  }
  verticesFile.close();

  // perform the queries
  irmq::RMQ rmq;
  rmq.initialise(euler_levels);

  std::string outFile = args.results_dir + "/lca.csv";
  std::ofstream out(outFile.c_str(), std::ios::out);

  #pragma omp parallel for shared(out, queries, euler_circuit, euler_positions, euler_levels, rmq) default(none)
  for (boost::uint32_t i = 0; i < (queries.size() - 1); ++i) {
    itypes::StringStringMap lcas;
    for (boost::uint32_t j = i + 1; j < queries.size(); ++j) {
      std::string left = euler_positions[queries[i]] < euler_positions[queries[j]] ? queries[i] : queries[j];
      std::string right = euler_positions[queries[j]] >= euler_positions[queries[i]] ? queries[j] : queries[i];
      std::string pair = left + "," + right;

#ifndef NDEBUG
      std::cout << "ID: " << left << " -> " << right << std::endl;
      std::cout << "Pos: " << euler_positions[left] << " -> " << euler_positions[right] << std::endl;
#endif /* NDEBUG */

      std::string lca = euler_circuit[rmq.query(euler_levels, 1, 0, euler_levels.size() -1, euler_positions[left], euler_positions[right])];
      lcas[pair] = lca;

#ifndef NDEBUG
      std::cout << pair << ": " << lca << std::endl;
#endif /* NDEBUG */
    }
    #pragma omp critical
    {
      itypes::StringStringMap::iterator it_begin = lcas.begin(), it_end = lcas.end();
      for (; it_begin != it_end; it_begin++) {
        out << it_begin->first << "," << it_begin->second << std::endl;
      }
    }
  }
  out.close();

  return EXIT_SUCCESS;
}


src/main/RMQException.cc
// CoPYRight (C) 2015 Dominik Dahlem <Dominik.Dahlem@gmail.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

#include "RMQException.hh"


namespace rmq
{


const char* RMQException::m_errorStrings[NUM_ERRORS] =
{
    "Error casting object.",
};

RMQException::RMQException(boost::uint8_t p_code) : exception(), m_code(p_code) {}

const char* RMQException::what() const throw() {
    return m_errorStrings[m_code];
}

boost::uint8_t RMQException::errorCode() const {
    return m_code;
}

}


src/main/CL.cc
// Copyright (C) 2015 Dominik Dahlem <Dominik.Dahlem@gmail.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

/** @file CL.cc
 * Implementation of the command-line parsing of the main routine
 *
 * @author Dominik Dahlem
 */
#if HAVE_CONFIG_H
# include <config.h>
#endif

#ifndef __STDC_CONSTANT_MACROS
# define __STDC_CONSTANT_MACROS
#endif /* __STDC_CONSTANT_MACROS */

#include <iostream>

#include <boost/cstdint.hpp>
#include <boost/filesystem.hpp>
namespace fs = boost::filesystem;

#include <boost/program_options/options_description.hpp>
#include <boost/program_options/parsers.hpp>
#include <boost/program_options/variables_map.hpp>
namespace po = boost::program_options;

#include "CL.hh"


CL::CL()
    : m_opt_desc(new po::options_description("Options"))
{
  // Declare the supported options.
  po::options_description opt_general("General Configuration");
  opt_general.add_options()
      (HELP.c_str(), "produce help message")
      (VERS.c_str(), "show the version")
      ;

  po::options_description opt_io("I/O Configuration");
  opt_io.add_options()
      (RESULTS_DIR.c_str(), po::value <std::string>()->default_value("./results"), "results directory.")
      (EULER_CIRCUIT.c_str(), po::value <std::string>()->default_value(""), "Filename of the Euler Circuit.")
      (EULER_LEVELS.c_str(), po::value <std::string>()->default_value(""), "Filename of the vertex levels in the Euler Circuit.")
      (EULER_POSITIONS.c_str(), po::value <std::string>()->default_value(""), "Filename of the vertex positions in the Euler Circuit.")
      (QUERY_SET.c_str(), po::value <std::string>()->default_value(""), "Filename of the query set to compute pair-wise LCAs over.")
      ;

  m_opt_desc->add(opt_general);
  m_opt_desc->add(opt_io);
}


int CL::parse(int argc, char *argv[], args_t &p_args)
{
  po::variables_map vm;

  po::store(po::parse_command_line(argc, argv, (*m_opt_desc.get())), vm);
  po::notify(vm);

  if (vm.count(HELP)) {
    std::cout << (*m_opt_desc.get()) << std::endl;
    return EXIT_FAILURE;
  }

  if (vm.count(VERS)) {
    std::cout << PACKAGE_NAME << " " << PACKAGE_VERSION << std::endl;
    std::cout << argv[0] << std::endl;
    return EXIT_FAILURE;
  }

  if (vm.count(RESULTS_DIR.c_str())) {
    p_args.results_dir = vm[RESULTS_DIR.c_str()].as <std::string>();
  }

  if (vm.count(EULER_CIRCUIT.c_str())) {
    p_args.euler_circuit = vm[EULER_CIRCUIT.c_str()].as <std::string>();
    if (!fs::exists(p_args.euler_circuit)) {
      std::cerr << "The filename " << p_args.euler_circuit << " with the Euler Circuit does not exist!" << std::endl;
      return EXIT_FAILURE;
    }
  }

  if (vm.count(EULER_LEVELS.c_str())) {
    p_args.euler_levels = vm[EULER_LEVELS.c_str()].as <std::string>();
    if (!fs::exists(p_args.euler_levels)) {
      std::cerr << "The filename " << p_args.euler_levels << " with the Vertex levels in the Euler Circuit does not exist!" << std::endl;
      return EXIT_FAILURE;
    }
  }

  if (vm.count(EULER_POSITIONS.c_str())) {
    p_args.euler_positions = vm[EULER_POSITIONS.c_str()].as <std::string>();
    if (!fs::exists(p_args.euler_levels)) {
      std::cerr << "The filename " << p_args.euler_positions << " with the Vertex positions in the Euler Circuit does not exist!" << std::endl;
      return EXIT_FAILURE;
    }
  }

  if (vm.count(QUERY_SET.c_str())) {
    p_args.query_set = vm[QUERY_SET.c_str()].as <std::string>();
    if (!fs::exists(p_args.euler_levels)) {
      std::cerr << "The filename " << p_args.query_set << " containing the query set does not exist!" << std::endl;
      return EXIT_FAILURE;
    }
  }

  std::cout << argv[0] << " " << PACKAGE_VERSION << std::endl;
  std::cout << PACKAGE_NAME << std::endl;
  std::cout << p_args << std::endl;

  return EXIT_SUCCESS;
}
